org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/ast/InstanceOfExpression.java

```diff
@@ -154,101 +149,64 @@ public void generateOptimizedBoolean(BlockScope currentScope, CodeStream codeStr
-		codeStream.dup();
 		codeStream.store(this.elementVariable.binding, false);
 	}
 
-	codeStream.load(this.secretInstanceOfPatternExpressionValue);
-	codeStream.removeVariable(this.secretInstanceOfPatternExpressionValue);
-	codeStream.checkcast(this.type, this.type.resolvedType, codeStream.position);
+	if (this.secretExpressionValue != null) {
+		codeStream.removeVariable(this.secretExpressionValue);
+	}
 
-	if (valueRequired && cst == Constant.NotAConstant) {
+	if (valueRequired) {
 		codeStream.generateImplicitConversion(this.implicitConversion);
 	} else {
 		codeStream.pop();
 	}
 	codeStream.recordPositionsFrom(pc, this.sourceStart);
 
-
-	if ((cst != Constant.NotAConstant) && (cst.typeID() == TypeIds.T_boolean)) {
-		pc = codeStream.position;
-		if (cst.booleanValue() == true) {
-			// constant == true
-			if (valueRequired) {
-				if (falseLabel == null) {
-					// implicit falling through the FALSE case
-					if (trueLabel != null) {
-						codeStream.goto_(trueLabel);
-					}
-				}
+	int position = codeStream.position;
+	if (valueRequired) {
+		if (falseLabel == null) {
+			if (trueLabel != null) {
+				// Implicit falling through the FALSE case
+				codeStream.goto_(trueLabel);
 			}
 		} else {
-			if (valueRequired) {
-				if (falseLabel != null) {
-					// implicit falling through the TRUE case
-					if (trueLabel == null) {
-						codeStream.goto_(falseLabel);
-					}
-				}
-			}
-		}
-		codeStream.recordPositionsFrom(pc, this.sourceStart);
-	} else  {
-		// branching
-		int position = codeStream.position;
-		if (valueRequired) {
-			if (falseLabel == null) {
-				if (trueLabel != null) {
-					// Implicit falling through the FALSE case
-					codeStream.pop2();
-					codeStream.goto_(trueLabel);
-				}
+			if (trueLabel == null) {
+				// Implicit falling through the TRUE case
 			} else {
-				if (trueLabel == null) {
-					// Implicit falling through the TRUE case
-					codeStream.pop2();
-				} else {
-					// No implicit fall through TRUE/FALSE --> should never occur
-				}
+				// No implicit fall through TRUE/FALSE --> should never occur
 			}
 		}
-		codeStream.recordPositionsFrom(position, this.sourceEnd);
 	}
+	codeStream.recordPositionsFrom(position, this.sourceEnd);
+
 	if (nextSibling != falseLabel)
 		nextSibling.place();
 }
 
-private void addAssignment(BlockScope currentScope, CodeStream codeStream, LocalVariableBinding local) {
-	assert local != null;
-	SingleNameReference lhs = new SingleNameReference(local.name, 0);
-	lhs.binding = local;
-	lhs.bits &= ~ASTNode.RestrictiveFlagMASK; // clear bits
-	lhs.bits |= Binding.LOCAL;
-	lhs.bits |= ASTNode.IsSecretYieldValueUsage;
-	((LocalVariableBinding) lhs.binding).markReferenced(); // TODO : Can be skipped?
-	Assignment assignment = new Assignment(lhs, this.expression, 0);
-	assignment.generateCode(currentScope, codeStream);
-	codeStream.addVariable(this.secretInstanceOfPatternExpressionValue);
-}
-
```

Comment:
This is too elaborate a dance. Just doing `codeStream.store(this.secretExpressionValue, true);` or `codeStream.store(this.secretExpressionValue, false);` suffices.